sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala

```diff
@@ -640,6 +639,26 @@ case class QualifyLocationWithWarehouse(catalog: SessionCatalog) extends Rule[Lo
   }
 }
 
+object TrimCollationCheck extends (LogicalPlan => Unit) {
```

Comment:
It's not enough for the statements like: `SELECT collate('aaa', 'UNICODE_TRIM')`.  Here default collation is UTF8_BINARY so changes in SQLConf won't help.

Reply:
What if we add the check also in the `CollateExpressionBuilder`?
https://github.com/apache/spark/blob/587cd2b18c37b35f641320eb176d7638b2e52b42/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collationExpressions.scala#L55

Reply:
Yes, that would work. Done!
For my own knowledge is there a reason we avoid adding analyzer rules?

Reply:
Analyzer is pretty slow right now, it runs the rules iteratively over and over again until the plan gets to the fixed point - which can be very slow for bigger queries. That's why I try to avoid introducing anything new that could increase that complexity further.

Btw, my analyzer knowledge is also very shallow, maybe @vladimirg-db can provide an expert opinion here 😄

Reply:
Sure.
- The rules run in fixed point (many iterations) over the whole logical plan, which can be super slow for big logical plans. Every new rule slows down those corner cases.
- Analyzer logic depends on the ordering of rules, which is very hard to get right, and there are frequent regressions related to unobvious outcome of some new rule.